### DISCUSSION TITLE
Windows App with ANE - App on non dev PC cannot find the ANE's dlls (AIR include paths)

### DISCUSSION THREAD

**Discussion by teotigraphix** [airsdk/Adobe-Runtime-Support#2727] [Q&A]
I have an application I wrote years ago that I could just copy the ane dlls to the root directory where the exe is for the main AIR application.

I am doing the same thing now and exporting a bundle app with windows ANE. Trying to run this on a completely separate PC I get the wonderful `3500 ArgumentError` error during the `ExtensionContextr.call()` method.

How does AIR look for dlls that are included with an ANE when the app is run as a bundle? I thought just putting them in the root would be fine but not here. 

The interesting part is I include these dlls in the AIR-SDK/bin directory when developing. Removing those files crash the app in the same way(obviously). So I know this has to be an include dll problem but I can't figure out how to get the application to run on a separate PC.

**Comment by ajwfrost** [comment]
On Windows, I would think the default search path for a DLL is within the same folder as the executable. But, when you specify a library as the entry point of the ANE, then AIR looks for this file in the expected subfolder location according to the normal installation or when packaged as a bundle.

One thing I would suggest checking for, if you're copying files over to a separate PC, is whether your DLL had any dependencies that may not be present on the new machine. Typically for example, if you are building a DLL using Visual Studio, you would need to have the appropriate Visual C runtime library present on the new machine. You can use "dependency walker" on the new PC to see whether there are any issues with the DLL (it seems to take a very long time to scan a DLL now, so you have to be patient!) - or check your linker settings and ensure it's all set to statically link everything?

To check further, we'd need to know whether you are deliberately requiring another DLL i.e. does your ANE have a DLL dependency that you need to distribute too? As this file may need to be put into the root of the application (i.e. where the .exe file is generated).

thanks

**Comment by marchbold** [comment]
Could be that you are relying on the c++ redistributable functionality but aren't installing it on the other machine.

https://docs.airnativeextensions.com/docs/nativewebview/add-the-extension#c-redistributable

**Comment by teotigraphix** [comment]
I tried the above, not working out. I have verified the dlls, must be some other Windows thing going on here. Thanks for the information @ajwfrost and @marchbold .

Seems like this is going to take a bit more investigation. 

The dll library is using fmodex.dll. Basically, there is the AIRNativeLib.dll(FRE bridge) then it loads AudioEngine.dll and that loads fmodex.dll. There is some other deps that have to do with windows, so I will check that but as far as I can see reading the dll, I have have the dependency dlls located and in the System32 directory.

**Reply by ajwfrost** [reply]
If you want, you could send us the package and we'll be able to add more tracing/debugging to see why the 3500 error code is coming out?

thanks

**Reply by teotigraphix** [reply]
Yeah that would be helpful, where should I send it?

**Reply by ajwfrost** [reply]
Please use https://transfer.harman.com/requests/9b1ipbZ8UnDJxhY0o4sVVQ
thanks

**Reply by teotigraphix** [reply]
Sent, thanks @ajwfrost .

**Reply by ajwfrost** [reply]
Received -> so when we expand the package and look at the DLLs in the ANE, using https://dependencywalker.com/, we can see that the core DLL that's mentioned in your extension descriptor file is a debug build .. it has dependencies upon "MSVCP140**D**.dll" and "VCRUNTIME140**D**.dll"

I think even if you rebuild that as a release build, you'd still then be dependent upon those Visual C++ redistributable packages. So either you can point people at https://www.microsoft.com/en-us/download/details.aspx?id=52685 or you can potentially set the builds up so that you statically link the necessary libraries: in Visual Studio project properties, go to the C/C++ section and find "Code Generation", there should be a "Runtime Library" entry that you can set to "Multi-threaded (/MT)" rather than "Multi-threaded DLL (/MD)".

thanks

**Reply by teotigraphix** [reply]
Oh man, a debug build, wow that escaped me, thanks for that. Makes perfect sense now.

Thanks for the other info, I will let you know, but this makes sense what is happening now.
